src/crypto/crypto_context.cc

```diff
@@ -188,28 +189,66 @@ int SSL_CTX_use_certificate_chain(SSL_CTX* ctx,
                                        issuer);
 }
 
+unsigned long LoadCertsFromFile(  // NOLINT(runtime/int)
+    std::vector<X509*>* certs,
+    const char* file) {
+  ClearErrorOnReturn clear_error_on_return;
+
+  BIOPointer bio(BIO_new_file(file, "r"));
+  if (!bio) return ERR_get_error();
+
+  while (X509* x509 = PEM_read_bio_X509(
+             bio.get(), nullptr, NoPasswordCallback, nullptr)) {
+    certs->push_back(x509);
+  }
+
+  unsigned long err = ERR_peek_error();  // NOLINT(runtime/int)
+  // Ignore error if its EOF/no start line found.
+  if (ERR_GET_LIB(err) == ERR_LIB_PEM &&
+      ERR_GET_REASON(err) == PEM_R_NO_START_LINE) {
+    return 0;
+  }
+
+  return err;
+}
+
 }  // namespace
 
-X509_STORE* NewRootCertStore() {
+X509_STORE* NewRootCertStore(Environment* env) {
   static std::vector<X509*> root_certs_vector;
+  static bool root_certs_vector_loaded = false;
   static Mutex root_certs_vector_mutex;
   Mutex::ScopedLock lock(root_certs_vector_mutex);
 
-  if (root_certs_vector.empty() &&
-      per_process::cli_options->ssl_openssl_cert_store == false) {
-    for (size_t i = 0; i < arraysize(root_certs); i++) {
-      X509* x509 =
-          PEM_read_bio_X509(NodeBIO::NewFixed(root_certs[i],
-                                              strlen(root_certs[i])).get(),
-                            nullptr,   // no re-use of X509 structure
-                            NoPasswordCallback,
-                            nullptr);  // no callback data
+  if (!root_certs_vector_loaded) {
+    if (per_process::cli_options->ssl_openssl_cert_store == false) {
+      for (size_t i = 0; i < arraysize(root_certs); i++) {
+        X509* x509 = PEM_read_bio_X509(
+            NodeBIO::NewFixed(root_certs[i], strlen(root_certs[i])).get(),
+            nullptr,  // no re-use of X509 structure
+            NoPasswordCallback,
+            nullptr);  // no callback data
+
+        // Parse errors from the built-in roots are fatal.
+        CHECK_NOT_NULL(x509);
 
-      // Parse errors from the built-in roots are fatal.
-      CHECK_NOT_NULL(x509);
+        root_certs_vector.push_back(x509);
+      }
+    }
 
-      root_certs_vector.push_back(x509);
+    if (!extra_root_certs_file.empty()) {
+      unsigned long err = LoadCertsFromFile(  // NOLINT(runtime/int)
+          &root_certs_vector,
+          extra_root_certs_file.c_str());
+      if (err) {
+        ProcessEmitWarning(env,
+                           "Ignoring extra certs from `%s`, load failed: %s\n",
+                           extra_root_certs_file.c_str(),
+                           ERR_error_string(err, nullptr));
+      }
```

Comment:
Don't use `ERR_error_string()`, it's not thread-safe. Use `ERR_error_string_n()` instead. (You probably copied that from `UseExtraCaCerts()`. It works there but it's still dubious.)

Don't call out to JS land with `root_certs_vector_mutex` locked. That deadlocks when  JS land calls back into `NewRootCertStore()`.

Calling out to JS land from here is questionable. Callers should probably handle that on a case-by-case basis. It's also somewhat questionable to turn failure into a warning instead of an error because failure affects the runtime behavior of the program.

Reply:
Agree 100% that it's dubious turning a failure into a warning here. The existing code did a `fprintf` during Node,js process startup with the text `"Warning: Ignoring extra certs from `%s`, load failed: %s\n`.

Switching to a failure would be a hard breaking change to Node.js. What do you recommend? I could switch it back to an `fprintf` but even with the original code that seemed a but dubious.

Reply:
I could live with a printf statement.

Reply:
Changes committed. Swapped to `ERR_error_string_n` and `fprintf`. Removed `Environment*` parameter from `NewRootCertStore()`.